### ISSUE TITLE
Email Settings Not being used for SMTP

### ISSUE THREAD

**Issue by tradedemon** [mzur/kirby-uniform#132]
Hi Mzur,

We've been scratching our head over this for a while and it seems the plugin isn't using the settings we've defined in the config.php.

_This is what I have in the kirby config.php (we're using SMTPgo as our SMTP provider)_

```php
// Define uniform actions
c::set('uniform.action', array(
  'service' => 'phpmailer',
  'service-options' => array(
    'host'     => 'mail.smtp2go.com',
    'port'     => 2525,
    'security' => 'tls',
    'user'     => 'username',
    'password' => '************'
  ),
  'from' => 'website@novi.io'
));
```

_This is what I have in my controller for that form._

```php
  $form->emailAction([
            // Send the notification to novi i.e. the receiver
            'to' => 'durranizee@gmail.com',
            'from' => 'website@novi.io',
            'subject' => 'Novi - New message from {name}',
            'snippet' => 'emails/contact-details',
        ])
        ->logAction([
            'file' => kirby()->roots()->site().'/messages.log'
        ])
```

1.) we aren't seeing any requests in our smtpgo.com dashboard indicating that messages are getting sent via the services

2.) This is a full email header that I am getting into my gmail account from the script and no reference to the SMTPgo service. It does however reference novi@novi.ie  which I haven't put in the config.php or the controller file for that form anywhere. Where is it getting that from ?

```
_Delivered-To: durranizee@gmail.com
Received: by 10.25.18.228 with SMTP id 97csp14014lfs;
        Thu, 22 Jun 2017 03:35:35 -0700 (PDT)
X-Received: by 10.80.132.99 with SMTP id 90mr2094361edp.50.1498127735686;
        Thu, 22 Jun 2017 03:35:35 -0700 (PDT)
ARC-Seal: i=1; a=rsa-sha256; t=1498127735; cv=none;
        d=google.com; s=arc-20160816;
        b=HnYgXZgEbIV2pDY90emw65t5drz72+JixL0jmCcjNrE1FgmRCCgUJYKlwWKPkeq2rb
         n/yiV/kG8VSdSmMm5LqXDyB1iW45lMp/kKH5S/wXigOcwZKsFp2KUrpfmwTytycgwV/r
         09E+bC3+GrKIKXK6b+stdCAyf+XJdAFxY82EkSbe2qiRuIdci3/ZF77TChxBsaLwZMBJ
         z8fMDZBISVwzDHIdN3V5RJb5pg41cRwZORSOsoeSzl/537zGNxfRRZ9YswBRLMPsxvkU
         TbQsQpc+Zza8X2XNFvaavLOliARAdLOuboi/hftZlADH0aSWARZwzTpLCwOlsF4JazQY
         5qLw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=google.com; s=arc-20160816;
        h=date:content-transfer-encoding:message-id:reply-to:from:subject:to
         :arc-authentication-results;
        bh=Lx0pWVkqTPHPls0DFJ9+JG0L0LutZ5QAJbHVQdlx5WU=;
        b=WPPkOemH0VvPUINLT5L8U5U/WLZ0ZH5pUMbUQ02VW8crKsFj2BLEpcKu+alyp6XRDp
         1kCNtZjTsCotGZVYK9vw7V0e+o1RyCwW8d1xJo3ayMmgKXe9V87D8M5D0URLHTezTg9F
         ejEzuXOzUmPe2s5jPiG6YACJ8xstZ90yK368R7IjpxTh0zGmiDIuhl7Iwc6hvzWpfAQt
         FmGoHu4wHWT5qCKUFoXZiKJ44hig/P4sgRAfpZNAesLOkgRTosm5LyLrXWCLt3lJnFEg
         rNsrCYtU6iQSTb2wqdMW95CZuDRP0lJQm3fa7Jl83zmh0yBxuceQH0cowMeIoKan7QV0
         hwlg==
ARC-Authentication-Results: i=1; mx.google.com;
       spf=fail (google.com: domain of novi@novi.ie does not designate 46.22.137.69 as permitted sender) smtp.mailfrom=novi@novi.ie
Return-Path: <novi@novi.ie>
Received: from novi.ie ([46.22.137.69])
        by mx.google.com with ESMTPS id p7si1285246edl.291.2017.06.22.03.35.35
        for <durranizee@gmail.com>
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Thu, 22 Jun 2017 03:35:35 -0700 (PDT)
Received-SPF: fail (google.com: domain of novi@novi.ie does not designate 46.22.137.69 as permitted sender) client-ip=46.22.137.69;
Authentication-Results: mx.google.com;
       spf=fail (google.com: domain of novi@novi.ie does not designate 46.22.137.69 as permitted sender) smtp.mailfrom=novi@novi.ie
Received: by novi.ie (Postfix, from userid 10000) id F40F01C17FB; Thu, 22 Jun 2017 11:35:34 +0100 (IST)
To: durranizee@gmail.com
Subject: Novi - New message from Daniel
X-PHP-Originating-Script: 10000:email.php
From: website@novi.io
Reply-To: durranizee@gmail.com
Message-ID: <1498127734-website@novi.io>
X-Mailer: PHP v7.0.19
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: 8bit
Date: Thu, 22 Jun 2017 11:35:34 +0100 (IST)

You have a new message.

Name: Daniel

Email: durranizee@gmail.com

Phone: 23432432432

Company: dd

Message: sddfdsfsdfsdfsd_
```

**Comment by mzur** [comment]
Where do you got that part with configuring the action in `config.php` from? Uniform never worked that way. You always have to supply the full options array to the action:

```php
$form->emailAction([
  'service' => 'phpmailer',
  'service-options' => array(
    'host'     => 'mail.smtp2go.com',
    'port'     => 2525,
    'security' => 'tls',
    'user'     => 'username',
    'password' => '************'
  ),
  'from' => 'website@novi.io'
   'to' => 'durranizee@gmail.com',
   'subject' => 'Novi - New message from {name}',
   'snippet' => 'emails/contact-details',
])
```

You can easily implement loading of the config value on your own, though. Check this out:


```php
$form->emailAction(array_merge(c::get('uniform.action'), [
   'to' => 'durranizee@gmail.com',
   'subject' => 'Novi - New message from {name}',
   'snippet' => 'emails/contact-details',
]))
```

**Comment by tradedemon** [comment]
Awesome dude, I'm not entirely sure. I picked this project up off from someone else and it was always in there. I assumed it was working until we ran some tests and noted it didn't work :)

**Comment by tradedemon** [comment]
I've changed all the settings as you've mentioned. I did however keep getting the error 

There was an error sending the form: The email service is not available: phpmailer

I have uploaded PHPMailer into the root public_html but it still doesn't work. I thought this might have to be something that has to be compiled into apache, but our hosting guys said that you just have to upload the files into the public_html. 

Can you clarify if we have any other options like maybe using the standard mailer on localhost or server.

**Comment by mzur** [comment]
You need to define a `phpmailer` email service. See [this post](https://forum.getkirby.com/t/contact-form-with-phpmailer-uniform/6469?u=mzur) for how to do that.

**Comment by tradedemon** [comment]
Ok thanks, I've done the above, now I just get the error.

"The email could not be sent."

Any ideas on how I can figure what's going on.

Thanks
Zee

**Comment by mzur** [comment]
That was actually a shortcoming of the email action which didn't pass through error messages from email services. I've just released a new version of Uniform where this is fixed. Be sure to `c::set('debug', true)` in the config so the error message will be shown with the new version.

This doesn't fix your error but the new error message will be helpful.

**Comment by tradedemon** [comment]
ok updated and am now getting 

_"There was an error sending the form: Undefined property: Email::$attachment"_

My form isn't attaching anything and I couldn't find anything in the docs that mentions an attachment property.

**Comment by mzur** [comment]
In this case, delete the following from the `phpmailer` email service:
```php
if ($email->attachment != null) {
   $mail->addAttachment($email->attachment);
}
```

**Comment by tradedemon** [comment]
ok removed that and it's fixed that issue.

**Comment by mzur** [comment]
And PHPMailer works as intended?

**Comment by creative9** [comment]
The good news is yes it's all working great with PHPMailer so the above fix sorted that issue.

I do however have another issue. If you have two forms on the same page it seems you get the same CSRF token for each of the forms ? Is that correct and if yes will that cause any problems ?

**Comment by mzur** [comment]
> Is that correct and if yes will that cause any problems?

Yes, that's correct and no it will not cause problems. According to OWASP it's sufficient that the token is [unique per user session](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Synchronizer_.28CSRF.29_Tokens).

**Comment by creative9** [comment]
Hmm that's strange, I've got a strange issue it's probably not your plugin but have any of your users ever come across this.

The site novi.ie has a link in the top menu which slides out a support form. When you fill in this form for the first time I get the INVALID CSRF token. But if you try the form again it works fine and it doesn't show that error. In fact except for the first time it works fine. It also seems to work fine from all other pages except for the homepage.

I took a look at the code to see if there might be duplicate calls to the homepage, or a dodgy image url  or something but I can't find anything. And if it was one of those issues why does it not work only on the first time.

I appreciate this might not be a plugin issue but has anyone else ever experienced this because I have no idea what's going on here

**Comment by mzur** [comment]
Does the form have any CSRF token when it's first loaded? That is, has the `csrf_token` field a value?

**Comment by tradedemon** [comment]
yes it does, the page is at.

http://novi.ie

I just did an inspect and can see this on the support form.

<input type="hidden" name="csrf_token"
value="VYYqwSKaOrG7573mJMrwDYG8xR7IXyNINJK6Z1Sy915uaMQ1PlwhNOa8YCXgsUGh">

Zee

**Comment by mzur** [comment]
You could hijack the `csrf` function of the kirby toolkit and see if it is called multiple times on the home page (it should be). Then you can use `debug_print_backtrace` to see where the second call came from.

**Comment by creative9** [comment]
Not to sure how to hijack, but I’m assuming it’s going into the helpers.php file and adding the debug_print_stacktrace to the function there ?

 

If that’s correct then I get an out of memory error.

 

The function is called twice though but on two separate forms so it should be ok right ?

 

Zee

**Comment by mzur** [comment]
> Not to sure how to hijack, but I’m assuming it’s going into the helpers.php file and adding the debug_print_stacktrace to the function there ?

That's exactly what I meant.

> The function is called twice though but on two separate forms so it should be ok right?

No, it should be called only once for each page. Do you use the `csrf` function directly anywhere or do you exclusively use the `csrf_field` helper?

**Comment by tradedemon** [comment]
No I just use the csrf_field but maybe it’s only called once. I haven’t verified that i only know that it’s included in two places on the same page and the token is exactly the same.

However as you stated before that shouldn’t be a problem right ?

**Comment by mzur** [comment]
> However as you stated before that shouldn’t be a problem right ?

Right. You can verify that `csrf` is called only once by "hijacking" the function with `var_dump` instead of the memory hungry `debug_print_backtrace`.

**Comment by creative9** [comment]
ok so I added the var_dump() function to the csrf_field like below. Is that correct it prints out twice on the page ?

function csrf_field($t = null)
    {
        // remember the token for multipme function calls
        static $token = null;
        $token = $token ?: csrf();
        // the token parameter overrides the generated token


        var_dump("Called csrf");
        return '<input type="hidden" name="'.Form::CSRF_FIELD.'" value="'.($t ?: $token).'">';
    }

**Comment by mzur** [comment]
That's correct. But try this with the `csrf` helper function of the Kirby Toolkit. This shouldn't be called twice.

**Comment by creative9** [comment]
Ok did the below and I don’t get any screen prints and the same error still happens.

```
function csrf($check = null) {

  // make sure a session is started
  s::start();

  var_dump("Called csrf");

  if(is_null($check)) {
    $token = str::random(64);
    s::set('csrf', $token);
    return $token;
  }
  return ($check === s::get('csrf')) ? true : false;
}
```

**Comment by mzur** [comment]
You probably have to search the HTML file for `"Called csrf"` because it may be printed where you can't see it.

**Comment by creative9** [comment]
Nope did a view source and search and nothing there.

Just to confirm the file I edited was kirby/toolkit/helpers.php

**Comment by mzur** [comment]
Okay, I find three forms in there. But all have the same CSRF token so this is working as it should (even though it's weird that the `var_dump` doesn't show up). Do you maybe have something that destroys/renews the session after the homepage is loaded? Other than that, I don't have a clue. The only thing you can do then is disable/comment out parts of the page and see when the form starts working again.

Another thing you can try is upgrading to Kirby 2.5. In this version the CSRF token is permanently stored in the session and survives page loads. But this way we'd still not know what went wrong in the first place.

**Comment by creative9** [comment]
Upgraded to 2.5 and still has the same problem.

I've removed a lot of the code but it still doesn't work. 

The only thing I've established through all this is 2 points.

1.) It only ever happens on the home page i.e. dev.novi.ie
2.) It always happens once then on subsequent attempts everything works fine.

These are the only 2 points which are consistent each time i test. Is there anything in the plugin that might check for something that invovles the above. Why would it only ever fail on the first attempt, that just seems really odd and why is it fine on all other pages i.e. dev.novi.ie/about works perfectly every time.

I'm really baffled now.

**Comment by creative9** [comment]
Surely if there's something in the HTML/JS code on the homepage then it shouldn't work on the second or third attempt. Is it possible that on the homepage they way it reads the token is affected i.e. there is a mismatch between whats on the page and what the code has ?

**Comment by mzur** [comment]
> I've removed a lot of the code but it still doesn't work.

Keep removing bit by bit until you have a blank page. Then just add a single form (without JS, CSS or anything) again and see if it works.

> It always happens once then on subsequent attempts everything works fine.

What do you do so it *doesn't* work again? Delete the cookies?

**Comment by creative9** [comment]
Ok will keep going.

If I do a CTRL+F5 (Deep refresh it stops working).
If I submit get the error then hit back, then submit it again it works.

**Comment by creative9** [comment]
I have stripped this sucka down to barebones and still exactly the same results.

You can see the page here:

http://dev.novi.ie

Let me know if you get the same problem or if you can suggest something further ?

Zee

**Comment by mzur** [comment]
Looks like it's working correctly for me. Try disabling your browser plugins or try another browser.

**Comment by creative9** [comment]
Yup everything is disabled and doesn't work in chrome or firefox.

IE i tested and it worked there. I assume you're using IE then ?

You need to clear your cache, cookies etc to make it throw that error.

**Comment by mzur** [comment]
> I assume you're using IE then ?

How dare you? :smile: No, I'm using Firefox 55 on Linux with a private window (without caching or cookies) and it works. Do you use a password manager or antivirus thingy that might intercept and modify the request?

**Comment by creative9** [comment]
haha thought that would give us a little chuckle. lol

I don't use password manager or antivirus that might do anything like that. Besides even if I did then the forms wouldn't work on any of the pages right ?

**Comment by mzur** [comment]
That's right. It's only that I'm running out of ideas. Do you consistently have this issue with your local development instance, too? Maybe try it on another machine?

**Comment by creative9** [comment]
No it's the first time. I'm getting some of the other guys at work to verify if they are getting the same issue too, will let you know what I find.

**Comment by creative9** [comment]
Man it's been a whole week of testing and working on this and we've made no progress. The page is bare bones there's no extra JS or anything funny in there and it still gives not only me the error but our testers.

I've also tried it on a number of devices and it's the same every time.

Are you sure you cleared the cache, opened a new tab for it to work ? The fact there's no way to debug this is also quite annoying.

**Comment by chbach** [comment]
Just fyi: I tested it on Chrome and Firefox for Mac and it just worked™ 

Both times, I got the 

```Thank You

A member of our team will be in touch shortly.
```

and no error.

**Comment by creative9** [comment]
Mind Boggling any ideas on how this can be debugged ?

**Comment by mzur** [comment]
You could ask the people in the Kirby forum to try your site. Maybe we can find a pattern if more people try it.
